### DISCUSSION TITLE
Synthetic dataset

### DISCUSSION THREAD

**Discussion by skirdey** [sensein/senselab#22] [Ideas]
Hi folks! Love this package, lots of useful features!

I wanted to propose a synthetic speech dataset that would include generated speech data of healthy controls - meaning speech that lacks any specific biomarkers of physical or mental health conditions.
The idea is that synthetic speech, sampled from models like XTTS2 or cloned from healthy individuals can serve as a great baseline for many experiments in voice biomarkers and digital phenotyping. 

I see the package already integrates with huggingface, it could be a perfect place to host synthetic speech dataset that could also serve power the unit tests. 

At Amplifier, we can provide GPU compute to get the first iteration of data generated and analyzed. 

The composition of the dataset is still up in the air, I was thinking something balanced across 4 age groups, biological sexes such as male, female and intersex, and languages starting with English and Spanish. 

We can provide analysis from our foundational biomarker model to gain confidence that synthetic speech lacks particular health condition signals, and I am sure you have pre-trained models that detect specific impairment that can also be used to eliminate samples that are on the border between healthy controls and tests. 

The value of synthetic speech dataset can be varied and helpful for researchers and startups tackling voice as vital sign problem or speech accessibility problems. 

Let me know what you folks think!

**Comment by fabiocat93** [comment]
Hi @skirdey, that’s a great idea! Thanks for sharing it! 

Here are some points to think about:

- _What Does "Healthy" Mean?_ In a general sense, what does it mean for speech to be "healthy"? In some studies targeting a specific pathology, it's easy to classify anything, not that pathology as the healthy control group. But here, since we don't have specific disorders or diseases in mind, are we talking about the absence of any physical or mental health indicators or something more specific? It may be important to define this clearly.

- People might want the healthy controls to match in terms of sentences, emotions, and other variables. Should we create a standard dataset with pre-defined samples or a flexible method/pipeline that allows users to generate a healthy control dataset on demand? Each approach has its pros and cons. What do you think?

- I see how the dataset can be validated by checking for some specific biomarkers. However, I would argue that there are several more aspects to consider. For instance, 
-- Do we want to verify the identity of the pseudo-speakers to ensure they don't resemble real individuals too closely and that utterances by the same pseudo-speaker can be consistently assigned to the same pseudo-speaker?
-- Should we test the intelligibility of the synthetic speech to make sure it's clear and understandable?
-- Do we want to check how the synthetic speech accurately conveys emotions?

I hope we can start shaping the direction of our project from yours and these points!

**Reply by skirdey** [reply]
Thanks @fabiocat93!

1 - absence of any physical or mental health indicators - I was thinking to eliminate any indicators to the best of our abilities, so we do not capture depressive episodes, or voice nodules, or diabetes or poor hydration, or something else. It is hard to say what is exactly healthy voice, but I guess we can say it is a voice that lacks detectable conditions that were verified by current voice biomarkers research. I think voice of Alexa or Siri most likely represents that extremely neutral "healthy" speech distribution. 

2 - Maybe a good starting point would be a standard dataset with LLM generated sentences that follow similar pattern as Rainbow passage that could expose different phonetic complexities. We could chunk the sentences into intervals of 5s, 10s, 20s, 30s - these lengths seem to be most common in the voice research, but I might be wrong. 

3 - Here we could automate some of it, such as transcription to make sure speech is intelligible, and maybe do the embedding similarity to make sure same pseudo-speaker maps to the same pseudo-speaker consistently across languages and sentences.  The anonymity part is probably need to be defined by voice licensing. If we could break the connection between source of the pseudo-speaker from the generated speech it would be great.

**Reply by skirdey** [reply]
And for the emotion part - that would be really cool to cover, at least some of the basic ones like angry, happy, sad, neutral

**Reply by fabiocat93** [reply]
> 1 - absence of any physical or mental health indicators - I was thinking to eliminate any indicators to the best of our abilities, so we do not capture depressive episodes, or voice nodules, or diabetes or poor hydration, or something else. It is hard to say what is exactly healthy voice, but I guess we can say it is a voice that lacks detectable conditions that were verified by current voice biomarkers research. I think voice of Alexa or Siri most likely represents that extremely neutral "healthy" speech distribution.


Sounds good. We can start from what we know. We should just clearly state our inclusion and exclusion criteria

> 2 - Maybe a good starting point would be a standard dataset with LLM generated sentences that follow similar pattern as Rainbow passage that could expose different phonetic complexities. We could chunk the sentences into intervals of 5s, 10s, 20s, 30s - these lengths seem to be most common in the voice research, but I might be wrong.


I like that. I don't know how much LLMs know about phonemes, but we may want to try to include and balance different sounds. 

> 3 - Here we could automate some of it, such as transcription to make sure speech is intelligible, and maybe do the embedding similarity to make sure same pseudo-speaker maps to the same pseudo-speaker consistently across languages and sentences. The anonymity part is probably need to be defined by voice licensing. If we could break the connection between source of the pseudo-speaker from the generated speech it would be great.
> And for the emotion part - that would be really cool to cover, at least some of the basic ones like angry, happy, sad, neutral


Agree. This is definitely feasible.

**Reply by skirdey** [reply]
Awesome, I think we can start with a set of 100 samples per unique combination of groups with documentation where the source samples are from, what was the license of that speech data. We can re-use the methodology you folks created to measure speaker similarity using embeddings. I saw quite a few speech quality metrics, but haven't tried them yet, so if you have one in mind to include we can also add it to the process. 

A few potential sources of license-friendly healthy control data for seniors and adults. 
https://zenodo.org/records/2867216
https://ieee-dataport.org/open-access/italian-parkinsons-voice-and-speech

**Reply by fabiocat93** [reply]
For assessing speech quality and intelligibility, we can use quantitative metrics such as:

- Wideband Perceptual Estimation of Speech Quality (PESQ)
- Short-Time Objective Intelligibility (STOI)
- Scale-Invariant Signal-to-Distortion Ratio (SI-SDR)

and qualitative metrics as:
- subjective Mean Opinion Score (MOS)

They are all implemented in [Torchaudio-Squim](https://arxiv.org/abs/2304.01448), and we can easily integrate in Senselab

**Reply by skirdey** [reply]
These are great. 

For synthetic speech generation, to make sure we are conform to Apache or MIT license, here are some projects that produce natural voices:


https://github.com/huggingface/parler-tts - Apache 2.0 License - great generic voices
https://github.com/metavoiceio/metavoice-src - Apache 2.0 License, supports voice cloning and quality is quite good 
https://github.com/myshell-ai/OpenVoice - MIT License
https://huggingface.co/suno/bark - MIT License
https://github.com/RVC-Boss/GPT-SoVITS - MIT License, supports voice cloning

**Reply by skirdey** [reply]
@fabiocat93 what do you think about age groups?

I thought to use approximately this mapping:
Child: 4-12 years
Adolescent: 13-19 years
Adult: 20-64 years
Senior: 65 years and older

**Reply by fabiocat93** [reply]
for TTS, I also like [tortoise-tts](https://github.com/neonbjb/tortoise-tts?tab=Apache-2.0-1-ov-file) under Apache 2.0 License, which supports voice cloning

**Reply by fabiocat93** [reply]
> @fabiocat93 what do you think about age groups?


also, children tend to undergo rapid changes in their voices due to their physical growth. Hence, it would be beneficial to have smaller age segments. But I see how we have to start somewhere, hence the group ages you suggested are ok for me

**Comment by satra** [comment]
Is the plan here to essentially use cloning of actual voices to create this dataset? or a generative model to create a dataset whose estimable parameters fall into the different groups under consideration?

**Reply by skirdey** [reply]
This project would focus on voice cloning to create the dataset, which can be a great segue towards a more universal TTS model that could produce healthy control speech data consistently.

**Reply by satra** [reply]
in that case, we absolutely need to dig into the rights and reuse of any voice that is cloned, and keep the ethics forefront in terms of potential uses of that voice.

**Reply by skirdey** [reply]
totally, there are quite a few datasets that provide data with established healthy controls and under MIT / Apache / Creative Commons licenses - I think it will be essential to keep the trace of where cloned voice came from and under what license in a form of a document as part of the dataset

**Reply by fabiocat93** [reply]
@skirdey just to clarify, when you say voice cloning, you mean a text-to-speech model able to emulate a specific target voice, don't you? (and not a speech-to-speech model, as for example FREEVC)

**Reply by skirdey** [reply]
I would say voice to voice like FREEVC as it often captures breathing and some other features unlike pure TTS, as capturing realistic patterns from original voice could provide stronger healthy control signal. 

The closest would be a system like VoiceCraft that accepts text + original voice sample

**Reply by satra** [reply]
i just played with voicecraft and it worked really nicely.

**Comment by skirdey** [comment]
I am going to do a short pass at a few examples using parler-tts and share a google drive. parler-tts seem to give consistently natural voices that are also fully synthetic and apache 2.0 licensed, but they are missing the air gasps and inhalations,, which is common miss in many other TTS systems. 

For the content of audios - we can have a similar approach of having tts read out Rainbow and North Wind And The Sun passage, and then LLM to act as a non-prompted text generator to get the unprompted speech. 

- https://www.york.ac.uk/media/languageandlinguistics/documents/currentstudents/linguisticsresources/Standardised-reading.pdf
- https://read.gov/aesop/143.html

**Reply by skirdey** [reply]
@satra @fabiocat93 https://drive.google.com/file/d/1TEy5QXIxFz7oR894zka_18baKoHAN16j/view?usp=sharing - a folder with a few synthetic audio files and metadata. Let me know what you think.

**Reply by skirdey** [reply]
And we can add Squim metrics to the metadata file for each recording, example:

Report for adult-female-freeform-synthetic-0.flac:
+--------+---------+
| Metric |   Value |
+--------+---------+
| STOI   |  0.9946 |
| PESQ   |  3.3025 |
| SI-SDR | 25.3140 |
| MOS    |  3.6959 |
+--------+---------+

**Reply by fabiocat93** [reply]
What is the difference between conditions `0` and `1`? 

Some comments:
- I can hear some background noise (that may be removed with some simple filter).
- The voices don't sound supernatural to me, especially for the pauses they do between one sentence and the other. Did you generate the entire story, or did you do it sentence by sentence?

**Reply by skirdey** [reply]
0 and 1 are just two different generations with different seed. 
Noise - we can probably remove it from the source before cloning as it seems models transfer that acoustic feature. 
The naturalness will vary from one voice cloning model to the other, and yes - the amount of text generated at once will also affect it. I this example XTTS2 was given full paragraph.

**Reply by skirdey** [reply]
I've tried a few OSS solutions to clone, and Metavoice-1B seem to be fastest and provides most natural sounding samples.  Plus it is fully Apache 2.0 licensed. Here is a sample from the same speaker https://drive.google.com/file/d/1GoILzryM7WzewmYSNTtV5EyWQUMOWEns/view?usp=sharing

**Reply by fabiocat93** [reply]
Thanks @skirdey. I find this audio better than the others but I still have some concerns in terms of how natural it sounds. 

One thought: do you think it's helpful to include single words and/or nonwords in the dataset? I say that because I suspect models may sound more natural to the human ear when they produce 1 word only. What do you think?

**Reply by satra** [reply]
are there any evaluation metrics for naturalness in different languages? this example sounded australian to me, but would it sound natural to an australian? are there distributional metrics on phoneme durations, etc.,. per language we could use, or would we have to do ethnographic evaluations?

**Reply by skirdey** [reply]
> Thanks @skirdey. I find this audio better than the others but I still have some concerns in terms of how natural it sounds.
> 
> One thought: do you think it's helpful to include single words and/or nonwords in the dataset? I say that because I suspect models may sound more natural to the human ear when they produce 1 word only. What do you think?

This I don't really know. I haven't seen any studies on how much predictive power within a single word. Using something like VAD the sample can be split into individual words as well.

**Reply by skirdey** [reply]
> are there any evaluation metrics for naturalness in different languages? this example sounded australian to me, but would it sound natural to an australian? are there distributional metrics on phoneme durations, etc.,. per language we could use, or would we have to do ethnographic evaluations?

good point, I would imagine that will require humans to evaluate global or regional accents. haven't seen a metric for this sort of eval.

**Reply by skirdey** [reply]
took another pass on generating a few more examples, using all healthy controls from the Parkinson's dataset - https://drive.google.com/file/d/1uvIgc5zVDtDa4RwZ9B37tHxdljm8nEhV/view?usp=sharing

It was generated sentence by sentence using cloning built-in into metavoice-1b. I like a lot about these samples, but there are still some artifacts that sound a bit un-natural. As next step, I am going to fine-tune metavoice on a few speakers and see if it helps reduce the artifacts. LMK what you think about structure and quality.

**Reply by fabiocat93** [reply]
Oh wow! I think this one is pretty accurate already. But I agree that it sounds noisy and with noisy I mean as if it was recorded through an old-school microphone

**Comment by fabiocat93** [comment]
hey @skirdey , do you know https://github.com/Camb-ai/MARS5-TTS? it sounds very clear, and I like the prosody. I think we should take it into account. It's under GNU AGPL 3.0 license, but they say we can request to use it under a different license by emailing them.

**Reply by skirdey** [reply]
MARS5 sounds really impressive, I am going to try it out on the same controls and share the results

**Reply by skirdey** [reply]
@fabiocat93 I've tried it on a few samples, and it seems to have a bug at the moment, when initial few seconds of speech are impressive, but then performance degrades quite a lot. Here is an example. It seems artifacts start at 12 seconds https://drive.google.com/file/d/1sfDT31to_SxOvC8rq5l6RA_DUmwxwrWQ/view?usp=drive_link

**Reply by skirdey** [reply]
after some tweaks and limiting generation to 12 seconds, I really like the result, we can totally use Mars. @fabiocat93 do you want to email their team and ask about the license? 

Here is an example of north sun passage split by sentence from a healthy control dataset clone - https://drive.google.com/file/d/1AbTr9BwPyKJdpkVv1K_lQ_6xODTGkloZ/view?usp=drive_link

**Reply by skirdey** [reply]
It maintains very natural breathing patterns and also produces audio in 24khz which is a nice bonus

**Reply by fabiocat93** [reply]
I am happy you liked it. I will email them and keep you all posted on what they say. Fingers crossed!

**Comment by fabiocat93** [comment]
Also, here is an interesting paper that we should take into account while designing our dataset: https://dl.acm.org/doi/pdf/10.1145/3593013.3594049
